### PR TITLE
Improve Core.Data Serialization and Add Remember Node

### PR DESCRIPTION
**Pull request by saintentropy**
### Purpose

The purpose of this PR is to improve the implementation of the Data.ParseJson and Data.StringifyJson to support json serialization of most Dynamo OOB data types.  Today these Data nodes fail to serialize or later to deserialize most Objects vs Value types.  Geometry for example always fails in StringifyJson.  This PR will add support for the following Data found in Dynamo OOB.

- ProtoGeometry types with the exception of Topology (as Topology does not make as much sence in isolation from the source) 
       - Primitive Types (BoundingBox, CoordinateSystem, Plane, Vector, Arc, Circle, Ellipse, EllipseArc, Line, NurbsCurve, PolyCurve, Polygon, Point, and NurbsSurface) are supposed via schema defined in [Autodesk Schema Service](https://schema.autodesk.com/home) 
       - Dynamo Geometry Types (Rectangle, Cuboid, Cone, Cylinder, Sphere, and Mesh) are supported via Dynamo specific schema.
       - Complex Geometry Types (PolySurface, Helix, Curves, Surfaces, Solids, TSplines) are supported via SAB and TSM

- Dynamo Core Types including Color, Location and Image (via png)

This PR also moves the implementation of the Remember node from the Generative Design package to Core.  The implementation of the Remember Node is identical but the serialization format has been updated to use Data.ParseJson and  Data.StringifyJson.  The version in GD will be marked as deprecated and removed from the package in future releases.  Due to the format change we are not migrating the node.

image

### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [x] This PR modifies some build requirements and the readme is updated
- [ ] This PR contains no files larger than 50 MB 

### Release Notes

Data.ParseJson and Data.StringifyJson nodes have been improved to support serialization of most Dynamo standard data types.   This includes all Geometry types except Topology and Color, Location and Image.

Data.Remember node has been moved from the GenerativeDesign package to Dynamo core library.  The new node has improved accuracy in how it stores and restores data.

### Reviewers

@sm6srw @QilongTang Please assign

### FYIs

@mjkkirschner @twastvedt @LongNguyenP

### CODE REVIEW

File: src/Libraries/CoreNodeModels/Remember.cs [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+using System.Text;
+using System.Threading.Tasks;
+using VMDataBridge;
+
+namespace CoreNodeModels
+{
+    [NodeName("Remember")]
+    [NodeDescription(nameof(Properties.Resources.RememberDescription), typeof(Properties.Resources))]
+    [NodeCategory("Core.Data")]
+    [InPortNames(">")]
```
Comment (aparajit-pratap) [inline src/Libraries/CoreNodeModels/Remember.cs:19]: ??

File: src/Libraries/CoreNodeModels/Remember.cs [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    [OutPortNames(">")]
+    [OutPortTypes("object")]
+    [OutPortDescriptions(typeof(Properties.Resources), nameof(Properties.Resources.RememberOuputToolTip))]
+    [IsDesignScriptCompatible]
+    [DynamoServices.RegisterForTrace]
+    public class Remember : NodeModel
+    {
+        private bool _deactivateNode;
+        private string _cache = "";
+        private string _updatedToolTipText = "";
```
Comment (aparajit-pratap) [inline src/Libraries/CoreNodeModels/Remember.cs:31]: We just use lower case names for private fields, no underscores.

File: src/Libraries/CoreNodeModels/Remember.cs (hunk 1 above)
Comment (saintentropy) [reply src/Libraries/CoreNodeModels/Remember.cs:19]: Node is designed to look like a watch node with ">" port strings

File: src/Libraries/CoreNodes/Data.cs [hunk 3]
```diff
@@ -51,16 +70,114 @@ private static object ToNative(JToken token)
...
+                        return CoordinateSystem.FromJson(jObject.ToString());
+
+                    //autodesk.geometry to abstract ProtoGeometry types
+                    case "autodesk.geometry:boundingbox3d-1.0.0":
+                        return BoundingBox.FromJson(jObject.ToString());
+
+                    case "dynamo.geometry:mesh-1.0.0":
+                        return Mesh.FromJson(jObject.ToString());
+
+                    //types supported by Goemetry.FromJson
```
Comment (sm6srw) [inline src/Libraries/CoreNodes/Data.cs:101]: spelling

File: src/Libraries/CoreNodes/Data.cs [hunk 4]
```diff
@@ -51,16 +70,114 @@ private static object ToNative(JToken token)
...
+                        return Geometry.FromJson(jObject.ToString());
+
+                    //Dynamo types
+                    case "dynamo.graphics:color-1.0.0":
+                        return Color.ByARGB(
+                        (int)jObject["A"],
+                        (int)jObject["R"],
+                        (int)jObject["G"],
+                        (int)jObject["B"]);
+#if _WINDOWS
```
Comment (sm6srw) [inline src/Libraries/CoreNodes/Data.cs:118]: Just curious, what is not supported in Linux here?

Comment (saintentropy) [reply src/Libraries/CoreNodes/Data.cs:118]: System.Drawing  The Bitmap aspect.  You also find similar depreciation for the Image nodes.  I don't know if I need to guard the code here but most was trying to play it safe.  The `JsonConverter`  get marked with
```
 #if NET6_0_OR_GREATER
         [SupportedOSPlatform("windows")]
 #endif 
```

File: src/Libraries/CoreNodeModels/Remember.cs [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+namespace CoreNodeModels
+{
+    [NodeName("Remember")]
+    [NodeDescription(nameof(Properties.Resources.RememberDescription), typeof(Properties.Resources))]
+    [NodeCategory("Core.Data")]
+    [InPortNames(">")]
+    [InPortTypes("object")]
+    [InPortDescriptions(typeof(Properties.Resources), nameof(Properties.Resources.RememberInputToolTip))]
+    [OutPortNames(">")]
+    [OutPortTypes("object")]
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodeModels/Remember.cs:19]: this should be `var` should it not?

File: src/Libraries/CoreNodeModels/Remember.cs [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+using System.ComponentModel;
+using VMDataBridge;
+
+namespace CoreNodeModels
+{
+    [NodeName("Remember")]
+    [NodeDescription(nameof(Properties.Resources.RememberDescription), typeof(Properties.Resources))]
+    [NodeCategory("Core.Data")]
+    [InPortNames(">")]
+    [InPortTypes("object")]
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodeModels/Remember.cs:16]: var?

File: src/Libraries/CoreNodes/CoreNodes.csproj [hunk 7]
```diff
@@ -40,6 +40,11 @@
       <Project>{c0d6dee5-5532-4345-9c66-4c00d7fdb8be}</Project>
       <Name>DesignScriptBuiltin</Name>
     </ProjectReference>
+    <ProjectReference Include="..\DynamoUnits\UnitsCore.csproj">
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodes/CoreNodes.csproj:43]: why is this necessary? I recall a lot of work untangling these dependencies.
I have not delved back in but this change might necessitate taking another look at the nuget packages.

File: src/Libraries/CoreNodes/CoreNodes.csproj [hunk 8]
```diff
@@ -40,6 +40,11 @@
...
+      <Project>{6e0a079e-85f1-45a1-ad5b-9855e4344809}</Project>
+      <Name>Units</Name>
+      <Private>False</Private>
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodes/CoreNodes.csproj:46]: also, does this duplicate the units dll?

File: src/Libraries/CoreNodes/Data.cs [hunk 9]
```diff
@@ -51,16 +70,114 @@ private static object ToNative(JToken token)
...
+                        return Vector.FromJson(jObject.ToString());
+
+                    case "autodesk.math:matrix44d-1.0.0":
+                        return CoordinateSystem.FromJson(jObject.ToString());
+
+                    //autodesk.geometry to abstract ProtoGeometry types
+                    case "autodesk.geometry:boundingbox3d-1.0.0":
+                        return BoundingBox.FromJson(jObject.ToString());
+
+                    case "dynamo.geometry:mesh-1.0.0":
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodes/Data.cs:98]: is this defined in the registry, or just in our code directly?

File: src/Libraries/CoreNodes/Data.cs [hunk 10]
```diff
@@ -51,16 +70,114 @@ private static object ToNative(JToken token)
...
+
+                    //Dynamo types
+                    case "dynamo.graphics:color-1.0.0":
+                        return Color.ByARGB(
+                        (int)jObject["A"],
+                        (int)jObject["R"],
+                        (int)jObject["G"],
+                        (int)jObject["B"]);
+#if _WINDOWS
+                    case "dynamo.graphics:png-1.0.0":
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodes/Data.cs:119]: there are no adsk image schemas?

File: src/Libraries/CoreNodes/Data.cs [hunk 11]
```diff
@@ -51,16 +70,114 @@ private static object ToNative(JToken token)
...
+
+                    default:
+                        return null;
+                }
+            }
+             
+            if (jObject.ContainsKey("typeid"))
+            {
+                var typeid = jObject["typeid"].ToString();
+                if (typeid == "autodesk.soliddef:model-1.0.0")
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodes/Data.cs:150]: till when will we support soliddef?

File: src/Libraries/CoreNodes/Data.cs [hunk 12]
```diff
@@ -51,16 +70,114 @@ private static object ToNative(JToken token)
...
+                    case "dynamo.geometry:sab-1.0.0":
+                    case "dynamo.geometry:tsm-1.0.0":
+                    case "dynamo.geometry:rectangle-1.0.0":
+                    case "dynamo.geometry:cuboid-1.0.0":
+                    case "dynamo.geometry:solid-1.0.0":
+                    case string geoId when geoId.Contains("autodesk.geometry"):
+                        return Geometry.FromJson(jObject.ToString());
+
+                    //Dynamo types
+                    case "dynamo.graphics:color-1.0.0":
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodes/Data.cs:112]: for all of these dynamo schemas, where do I find them?

File: src/Libraries/CoreNodes/Data.cs [hunk 13]
```diff
@@ -103,5 +220,226 @@ public override bool CanConvert(Type objectType)
...
+            public override bool CanConvert(Type objectType)
+            {
+                return typeof(DynamoUnits.Location) == objectType;
+            }
+        }
+
+#if NET6_0_OR_GREATER
+        [SupportedOSPlatform("windows")]
+#endif
+        private class ImageConverter : JsonConverter
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodes/Data.cs:328]: perhaps this should be called PNGImageConverter?

File: src/Libraries/CoreNodeModels/Remember.cs (hunk 6 above)
Comment (saintentropy) [reply src/Libraries/CoreNodeModels/Remember.cs:16]: This node does not replicate.  We can change the label to whatever we would like.  In the Node model case the port names are just for display.  What do you think is the oppriate label @mjkkirschner

File: src/Libraries/CoreNodes/CoreNodes.csproj (hunk 7 above)
Comment (saintentropy) [reply src/Libraries/CoreNodes/CoreNodes.csproj:43]: In terms of Nuget I think we are good.  CoreNodes (ie DSCoreNode.dll) is in the DynamoVisualProgramming.DynamoCoreNodes nuget.  That nuget depends on DynamoVisualProgramming.ZeroTouchLibrary nuget which includes DynamoUnits.dll aka UnitsCore.csproj.  That being said the purpose of this is to make it so we can serialize `Dynamo.Location` objects.  Not sure how important that one is

File: src/Libraries/CoreNodes/CoreNodes.csproj (hunk 8 above)
Comment (saintentropy) [reply src/Libraries/CoreNodes/CoreNodes.csproj:46]: Do you mean copy local?  Not sure what you are asking.

File: src/Libraries/CoreNodes/Data.cs (hunk 9 above)
Comment (saintentropy) [reply src/Libraries/CoreNodes/Data.cs:98]: Our code directly until we get it registered.  That process takes a bit.

File: src/Libraries/CoreNodes/Data.cs (hunk 10 above)
Comment (saintentropy) [reply src/Libraries/CoreNodes/Data.cs:119]: Not that I have found.

Comment (saintentropy) [reply src/Libraries/CoreNodes/Data.cs:119]: none in production... none in staging.

File: src/Libraries/CoreNodes/Data.cs (hunk 11 above)
Comment (saintentropy) [reply src/Libraries/CoreNodes/Data.cs:150]: I would say until we remove our nodes from LibG

File: src/Libraries/CoreNodes/CoreNodes.csproj (hunk 8 above)
Comment (mjkkirschner) [reply src/Libraries/CoreNodes/CoreNodes.csproj:46]: yeah I want to make sure we are not adding extra binaries to the nodes folder etc.

File: src/Libraries/CoreNodes/Data.cs (hunk 9 above)
Comment (mjkkirschner) [reply src/Libraries/CoreNodes/Data.cs:98]: is there a task somewhere for it?

File: src/Libraries/CoreNodeModels/Remember.cs (hunk 6 above)
Comment (mjkkirschner) [reply src/Libraries/CoreNodeModels/Remember.cs:16]: I think the appropriate label for a none replicating node is `var[]..[]`

File: src/Libraries/CoreNodeModels/Remember.cs [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+    [OutPortNames(">")]
+    [OutPortTypes("object")]
+    [OutPortDescriptions(typeof(Properties.Resources), nameof(Properties.Resources.RememberOuputToolTip))]
+    [IsDesignScriptCompatible]
+    [DynamoServices.RegisterForTrace]
+    public class Remember : NodeModel
+    {
+        private string cache = "";
+
+        [JsonProperty("Cache")]
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodeModels/Remember.cs:27]: why do you need this when the property has the same name?

File: src/Libraries/CoreNodeModels/Remember.cs [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+        }
+
+        public override void Dispose()
+        {
+            PropertyChanged -= OnPropertyChanged;
+            base.Dispose();
+            DataBridge.Instance.UnregisterCallback(GUID.ToString());
+        }
+
+        private static readonly string BuiltinDictionaryTypeName = typeof(DesignScript.Builtin.Dictionary).FullName;
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodeModels/Remember.cs:84]: can't you use nameof here?

File: src/Libraries/CoreNodeModels/Remember.cs [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+        {
+            RegisterAllPorts();
+            PropertyChanged += OnPropertyChanged;
+        }
+
+        private void OnPropertyChanged(object sender, PropertyChangedEventArgs e)
+        {
+            switch (e.PropertyName)
+            {
+                case "State":
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodeModels/Remember.cs:58]: nameof?

File: src/Libraries/CoreNodes/Data.cs (hunk 12 above)
Comment (mjkkirschner) [reply src/Libraries/CoreNodes/Data.cs:112]: do you think it makes sense to provide these schemas in the repo?

File: src/Libraries/CoreNodes/Data.cs [hunk 17]
```diff
@@ -51,16 +70,114 @@ private static object ToNative(JToken token)
...
+                        (double)jObject["Latitude"],
+                        (double)jObject["Longitude"],
+                        (string)jObject["Name"]);
+
+                    default:
+                        return null;
+                }
+            }
+             
+            if (jObject.ContainsKey("typeid"))
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodes/Data.cs:147]: whats the difference between serializations that contain "typeid" and "$typeid" ?

File: src/Libraries/CoreNodes/Data.cs [hunk 18]
```diff
@@ -51,16 +70,114 @@ private static object ToNative(JToken token)
...
+
+                            return bitmap;
+                        }
+#endif
+
+                        return null;
+
+                    case "dynamo.data:location-1.0.0":
+                        return DynamoUnits.Location.ByLatitudeAndLongitude(
+                        (double)jObject["Latitude"],
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodes/Data.cs:138]: curious - what are your thoughts about this vs adding protected or internal json constructors and just letting newtonsoft figure it out? 

I'm fine with this for now, just wanted to evaluate the +/-s

File: test/DynamoCoreTests/DSCoreDataTests.cs [hunk 19]
```diff
@@ -186,5 +191,359 @@ public void ParsingJSONInPythonReturnsSameResult()
...
+        public void RoundTripForBoundingBoxReturnsSameResult()
+        {
+            // Load test graph
+            string path = Path.Combine(TestDirectory, @"core\json\Abstract_BoundingBox_JSONParsing.dyn");
+            OpenModel(path);
+
+            // Verify objects match when serializing / de-serializing geometry type
+            AssertPreviewValue("abb39e07-db08-45cf-9438-478defffbf68", true);
+
+            // Verify current failure cases are all false
```
Comment (mjkkirschner) [inline test/DynamoCoreTests/DSCoreDataTests.cs:206]: I don't understand this comment.

File: test/DynamoCoreTests/DSCoreDataTests.cs [hunk 20]
```diff
@@ -186,5 +191,359 @@ public void ParsingJSONInPythonReturnsSameResult()
...
+            // Load test graph
+            string path = Path.Combine(TestDirectory, @"core\json\Point_JSONParsing.dyn");
+            OpenModel(path);
+
+            // Verify objects match when serializing / de-serializing geometry type
+            AssertPreviewValue("71efc8c5c0c74189901707c30e6d5903", true);
+        }
+
+        [Test]
+        [Category("Failure")]
```
Comment (mjkkirschner) [inline test/DynamoCoreTests/DSCoreDataTests.cs:388]: why does this fail? Can you add a TODO, task number etc?

File: test/DynamoCoreTests/DSCoreDataTests.cs [hunk 21]
```diff
@@ -186,5 +191,359 @@ public void ParsingJSONInPythonReturnsSameResult()
...
+            // Load test graph
+            string path = Path.Combine(TestDirectory, @"core\json\Curve_NurbsCurve_JSONParsing.dyn");
+            OpenModel(path);
+
+            // Verify objects match when serializing / de-serializing geometry type
+            AssertPreviewValue("423356e2c8f84e00aa6c50e9bdb72c98", true);
+        }
+
+        [Test]
+        [Category("Failure")]
```
Comment (mjkkirschner) [inline test/DynamoCoreTests/DSCoreDataTests.cs:340]: why does this fail - etc

File: test/DynamoCoreTests/DSCoreDataTests.cs [hunk 22]
```diff
@@ -186,5 +191,359 @@ public void ParsingJSONInPythonReturnsSameResult()
...
+            // Load test graph
+            string path = Path.Combine(TestDirectory, @"core\json\Abstract_Vector_JSONParsing.dyn");
+            OpenModel(path);
+
+            // Verify objects match when serializing / de-serializing geometry type
+            AssertPreviewValue("71efc8c5c0c74189901707c30e6d5903", true);
+        }
+
+        [Test]
+        [Category("Failure")]
```
Comment (mjkkirschner) [inline test/DynamoCoreTests/DSCoreDataTests.cs:247]: why does this fail

File: test/DynamoCoreTests/DSCoreDataTests.cs [hunk 23]
```diff
@@ -186,5 +191,359 @@ public void ParsingJSONInPythonReturnsSameResult()
...
+
+            // Verify objects match when serializing / de-serializing geometry type
+            AssertPreviewValue("07366adaf0954529b1ed39b240192c96", true);
+        }
+
+        [Test]
+        [Category("UnitTests")]
+        public void RoundTripForColorReturnsSameResult()
+        {
+            var color = DSCore.Color.ByARGB(25, 30, 35, 40);
```
Comment (mjkkirschner) [inline test/DynamoCoreTests/DSCoreDataTests.cs:403]: Is there any benefit to the preceding tests being implemented using graphs instead of implementing them to how these tests are written without involving graph deserialization? I find them easier to debug.

File: src/Libraries/CoreNodes/Data.cs [hunk 24]
```diff
@@ -103,5 +220,226 @@ public override bool CanConvert(Type objectType)
...
+        /// <summary>
+        /// Helper function to determine if object can be cached or if it is null, "null" string, or empty list.  
+        /// </summary>
+        /// <param name="inputObject">Object to check</param>
+        /// <returns></returns>
+        [IsVisibleInDynamoLibrary(false)]
+        public static bool CanObjectBeCached(object inputObject)
+        {
+            if (inputObject == null
+                || (inputObject is string inputString && inputString == "null")
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodes/Data.cs:376]: when do we have a "null" string?

Comment (mjkkirschner) [reply src/Libraries/CoreNodes/Data.cs:376]: what about an empty string?

File: test/DynamoCoreTests/DSCoreDataTests.cs [hunk 25]
```diff
@@ -186,5 +191,359 @@ public void ParsingJSONInPythonReturnsSameResult()
...
+        [Category("UnitTests")]
+        public void CanObjectBeCachedRejectsNullString()
+        {
+            var canCacheStringNull = DSCore.Data.CanObjectBeCached("null");
+            Assert.IsFalse(canCacheStringNull);
+        }
+
+        [Test]
+        [Category("UnitTests")]
+        public void CanObjectBeCachedRejectsEmptyString()
```
Comment (mjkkirschner) [inline test/DynamoCoreTests/DSCoreDataTests.cs:461]: this test seems to have the wrong name.

File: test/Tools/docGeneratorTestFiles/TestMdOutput_CoreNodeModels/CoreNodeModels.Remember.md [hunk 26]
```diff
@@ -0,0 +1,4 @@
+## Remember - Documentation
+This documentation file is auto generated by NodeDocumentationMarkdownGenerator, Version=2.13.0.2212, Culture=neutral, PublicKeyToken=null.
```
Comment (mjkkirschner) [inline test/Tools/docGeneratorTestFiles/TestMdOutput_CoreNodeModels/CoreNodeModels.Remember.md:2]: this seems like a great node to supply better documentation for.
FYI @hwahlstrom

File: src/Libraries/CoreNodes/Properties/Resources.resx [hunk 27]
```diff
@@ -141,6 +141,12 @@
   <data name="EnumDateOfWeekWednesday" xml:space="preserve">
     <value>Wednesday</value>
   </data>
+  <data name="Exception_Deserialize_Unsupported_Type" xml:space="preserve">
+    <value>The stored data can not be rebuilt.</value>
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodes/Properties/Resources.resx:145]: loaded?

File: src/Libraries/CoreNodes/Data.cs [hunk 28]
```diff
@@ -103,5 +220,226 @@ public override bool CanConvert(Type objectType)
...
+
+            //Try to serialize the inputs and return
+            string newCachedJson;
+            try
+            {
+                newCachedJson = StringifyJSON(inputObject);
+            }
+            catch(Exception ex)
+            {
+                throw new NotSupportedException(string.Format(Properties.Resources.Exception_Serialize_Unsupported_Type, inputObject.GetType().FullName));
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodes/Data.cs:433]: would it be helpful to also log the real exception to the console or concat it?

File: src/Libraries/CoreNodes/Data.cs [hunk 29]
```diff
@@ -103,5 +220,226 @@ public override bool CanConvert(Type objectType)
...
+                get { return false; }
+            }
+
+            public override bool CanConvert(Type objectType)
+            {
+                return typeof(DesignScriptEntity).IsAssignableFrom(objectType);
+            }
+        }
+
+        private class ColorConveter : JsonConverter
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodes/Data.cs:273]: why use a bunch of converters for this instead of decorating the class with attributes. I guess I get a bit of a code smell when the serialization and deserialization are handled in different places, one using a converter and one using regular constructors. Can it be unified?

File: src/Libraries/CoreNodeModels/Remember.cs [hunk 30]
```diff
@@ -0,0 +1,126 @@
...
+            //First build the function call
+            var functionCall = AstFactory.BuildFunctionCall(
+               new Func<object, string, Dictionary<string, object>>(DSCore.Data.Remember), funtionInputs);
+
+            var functionCallIndent = AstFactory.BuildIdentifier(GUID + "_func");
+
+            resultAst.Add(AstFactory.BuildAssignment(functionCallIndent, functionCall));
+
+            //Next add the first key value pair to the output port
+            var getFirstKey = AstFactory.BuildFunctionCall(BuiltinDictionaryTypeName, BuiltinDictionaryGet,
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodeModels/Remember.cs:102]: I think the ASTFactory somewhere has methods for building these calls easily.

File: src/Libraries/CoreNodeModels/Remember.cs [hunk 31]
```diff
@@ -0,0 +1,126 @@
...
+    [NodeDescription(nameof(Properties.Resources.RememberDescription), typeof(Properties.Resources))]
+    [NodeCategory("Core.Data")]
+    [InPortNames(">")]
+    [InPortTypes("object")]
+    [InPortDescriptions(typeof(Properties.Resources), nameof(Properties.Resources.RememberInputToolTip))]
+    [OutPortNames(">")]
+    [OutPortTypes("object")]
+    [OutPortDescriptions(typeof(Properties.Resources), nameof(Properties.Resources.RememberOuputToolTip))]
+    [IsDesignScriptCompatible]
+    [DynamoServices.RegisterForTrace]
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodeModels/Remember.cs:22]: how is trace used?

File: src/Libraries/CoreNodes/Data.cs (hunk 9 above)
Comment (saintentropy) [reply src/Libraries/CoreNodes/Data.cs:98]: Yes in the Vera backlog.  A few things we are following up on with the forge schemas.  1) Adding our Dynamo types where it makes senses. and 2) Adding support to the geometry schemas for things we need and make sense for a broader audience... for oriented bounding box, Rectangles, Cuboid Surfaces...  The mesh one is bit unclear.  There isn't an open source definition that anyone can agree on.  TrittonMesh is not open now.  We have a task to support glt as the standard mesh type but that is also complicated.  I think the main pattern is that we will 1) document these and 2) begin to introduce non dynamo specific schemas over time.  In those cases we will continue to deserialize these types but serialize the new types.

File: src/Libraries/CoreNodes/Data.cs (hunk 12 above)
Comment (saintentropy) [reply src/Libraries/CoreNodes/Data.cs:112]: I don't think we need the schemas here yet.  But also I don't think we should break.

File: src/Libraries/CoreNodes/Data.cs (hunk 17 above)
Comment (saintentropy) [reply src/Libraries/CoreNodes/Data.cs:147]: In ASP json that us `$typeid` as reserved keyword for the type.  In the changeset world they used `typeid`

File: src/Libraries/CoreNodes/Data.cs (hunk 18 above)
Comment (saintentropy) [reply src/Libraries/CoreNodes/Data.cs:138]: I think that is better approach.  I was waiting for the .NET7 so we could introduce an interface to support serialization registration similar to the IGraphicItem.  (static abstract is now supported in interfaces NET7 and above).  I can do this now without the interface for the internal types just to set the pattern.

File: test/DynamoCoreTests/DSCoreDataTests.cs (hunk 23 above)
Comment (saintentropy) [reply test/DynamoCoreTests/DSCoreDataTests.cs:403]: Yeah... Sorry I was just following the pattern of the other tests for Json.  Also added benefit of much greater coverage.  We can create a follow up task and pull all the tests into unit test style if it is critical

File: src/Libraries/CoreNodes/Data.cs (hunk 29 above)
Comment (saintentropy) [reply src/Libraries/CoreNodes/Data.cs:273]: I think the main reason is the overlap with the node -> file serialization and the node object -> data serialization.  I hope to introduce the interface so that it can be explicit.

File: src/Libraries/CoreNodeModels/Remember.cs (hunk 31 above)
Comment (saintentropy) [reply src/Libraries/CoreNodeModels/Remember.cs:22]: oops.  legacy.  Should not be

File: src/Libraries/CoreNodeModels/Remember.cs (hunk 14 above)
Comment (saintentropy) [reply src/Libraries/CoreNodeModels/Remember.cs:27]: Good catch

File: src/Libraries/CoreNodeModels/Remember.cs (hunk 15 above)
Comment (saintentropy) [reply src/Libraries/CoreNodeModels/Remember.cs:84]: It is shorted to "Dictionary"

File: src/Libraries/CoreNodeModels/Remember.cs (hunk 30 above)
Comment (saintentropy) [reply src/Libraries/CoreNodeModels/Remember.cs:102]: Not that I am aware of or could find.  The main thing we do here is that we need to pass the data from the inputPort and the cached string into and back out of the node every time.  The dictionary is used so that we can only present to the ports data on the node and keep the cache data internal.  I think we look long and hard at the time we wrote this and didn't have any out of the box options of dictionaries in ASTFactory.

File: src/Libraries/CoreNodes/CoreNodes.csproj (hunk 8 above)
Comment (saintentropy) [reply src/Libraries/CoreNodes/CoreNodes.csproj:46]: Should be good
image

File: test/DynamoCoreTests/DSCoreDataTests.cs (hunk 23 above)
Comment (mjkkirschner) [reply test/DynamoCoreTests/DSCoreDataTests.cs:403]: nah, not critical, maybe just preferred for future tests.

File: src/Libraries/CoreNodeModels/Remember.cs (hunk 30 above)
Comment (mjkkirschner) [reply src/Libraries/CoreNodeModels/Remember.cs:102]: look for:
DictionaryExpressionBuilder

Comment (mjkkirschner) [reply src/Libraries/CoreNodeModels/Remember.cs:102]: https://github.com/DynamoDS/Dynamo/blob/master/src/Engine/ProtoCore/Parser/DictionaryBuilder.cs

File: src/Libraries/CoreNodes/Data.cs (hunk 18 above)
Comment (mjkkirschner) [reply src/Libraries/CoreNodes/Data.cs:138]: ah I see, so we'll have to wait on the interface approach until after we move to .net 8... I guess you could do it now and change the language target of this csproj to 11?

File: src/Libraries/CoreNodes/Data.cs (hunk 12 above)
Comment (mjkkirschner) [reply src/Libraries/CoreNodes/Data.cs:112]: do you the tests cover all schemas?

Comment (saintentropy) [reply src/Libraries/CoreNodes/Data.cs:112]: Yes

File: src/Libraries/CoreNodes/Data.cs (hunk 29 above)
Comment (saintentropy) [reply src/Libraries/CoreNodes/Data.cs:273]: Nevermind.... file stuff doesn't apply to ZT.  But would be handled by the interface

File: test/DynamoCoreTests/DSCoreDataTests.cs (hunk 22 above)
Comment (saintentropy) [reply test/DynamoCoreTests/DSCoreDataTests.cs:247]: Failures are waiting for the upstream PR in LibG

File: src/Libraries/CoreNodeModels/Remember.cs (hunk 30 above)
Comment (saintentropy) [reply src/Libraries/CoreNodeModels/Remember.cs:102]: Ahhh.  Thanks.  I did look at that.  Interesting but not exactly what we need.  The dictionary is already set up by the function we call `Remember()` as the return type.  The AST construction we are doing is to look up specific keys later...  More specifically `DictionaryBuilder` embeds the constuctor `DesignScript.Builtin.Dictionary.ByKeysValues` whereas we  need the method `DesignScript.Builtin.Dictionary.ValueAtKey`.   There could be a helper for us but not this one.

**Review by mjkkirschner — APPROVED** [review]
@saintentropy lets ship it.

**Comment by saintentropy** [comment]
Cool... Will merge when CICD confirms

**Comment by saintentropy** [comment]
OK Passed on https://master-15.jenkins.autodesk.com/view/DYN/job/DYN-DevCI_Self_Service/1244/.  Going to merge 😃
